sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js

```diff
@@ -458,8 +468,276 @@ Ext.define('SlateTasksTeacher.controller.Dashboard', {
         });
     },
 
+    onAddGoogleAttachmentClick: function() {
+        var me = this,
+            taskEditor = me.getTaskEditor(),
+            token = Ext.util.Cookies.get('googleAppsToken', '/');
+
+        if (!taskEditor.getTask().phantom) {
+            Ext.Msg.alert('Error', 'Unable to add attachments while editing&hellip; Please create a new task.');
+            return;
+        }
+        // check for google token in cookies
+        gapi.load('auth:picker:client', {
+            callback: function() {
+                // TODO: check if token is any good
+                if (token) {
+                    return me.doOpenFilePicker(token);
+                }
+
+                me.doGoogleAuthentication();
+            }
+        });
+    },
+
+    onGoogleAuthentication: function(authResult) {
+        var me = this,
+            googleAppsDomain = SiteEnvironment && SiteEnvironment.googleAppsDomain || location.host,
+            googleAppsEmailRegex = new RegExp('^.+\@'+googleAppsDomain+'$'),
+            tokenExpiration,
+            _verifyUserIsWithinDomain, _userIsNotWithinDomain;
+
+        console.info('handleAuthResult(%o)', authResult);
+
+        if (!authResult || authResult.error) {
+            Ext.Msg.alert('Error', 'Unable to authenticate user. Please try again or contact an administrator.');
+            return;
+        }
+
+        _userIsNotWithinDomain = function(errorMessage) {
+            Ext.Msg.alert('Error', errorMessage);
+            return;
+        };
+
+        _verifyUserIsWithinDomain = function(response) {
+            console.info('loaded user', response.result);
+
+            if (!response.result || !response.result.user || !response.result.user.emailAddress.match(googleAppsEmailRegex)) {
+                return _userIsNotWithinDomain('Account must be under the correct domain. ' + googleAppsDomain);
+                // return false;
+            }
+
+            tokenExpiration = new Date(authResult.expires_at * 1000);
+
+            Ext.util.Cookies.set('googleAppsToken', authResult.access_token, tokenExpiration, '/');
+            Ext.util.Cookies.set('googleAppsPermissionId', response.result.user.permissionId, tokenExpiration, '/');
+
+            me.doOpenFilePicker(authResult.access_token);
+        };
+
+        // confirm authentication is within google apps domain
+        gapi.client.request({
+            path: '/drive/v3/about',
+            params: {
+                fields: 'user'
+            }
+        }).
+        then(_verifyUserIsWithinDomain);
```

Comment:
@nbey inline the function instead of storing it in a variable to keep things easier to follow unless there are multiple paths to that function that necessitate making it reusable